pkg/controller/hvpa/hvpa_controller.go

```diff
@@ -89,17 +162,145 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+			// Get latest HVPA object
+			hvpa := &autoscalingv1alpha1.Hvpa{}
+			err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: a.Meta.GetNamespace()}, hvpa)
+			if err != nil {
+				log.Error(err, "Error retreiving hvpa")
+				return nil
+			}
+			clone := hvpa.DeepCopy()
+
+			hvpaStatus := clone.Status.HvpaStatus
+			if hvpaStatus.LastScaleTime == nil || hvpaStatus.OverrideScaleUpStabilization == true {
+				log.V(4).Info("HVPA status already set to override last scale time.")
+				return nil
+			}
+
+			if pod.Status.Reason == "Evicted" {
+				// If pod was evicted beause of 'KubeletHasInsufficientMemory' node condition,
+				// only then we want to continue, otherwise exit
+				req := types.NamespacedName{
+					Name: nodeName,
+				}
+				node := corev1.Node{}
+				err := client.Get(context.TODO(), req, &node)
+				if err != nil {
+					log.Error(err, "Error fetching node")
+					return nil
+				}
+
+				hasMemPressure := false
+				for _, condition := range node.Status.Conditions {
+					if condition.Reason == "KubeletHasInsufficientMemory" && condition.Status == corev1.ConditionTrue {
+						hasMemPressure = true
+						break
+					}
+				}
+				if !hasMemPressure {
+					log.V(4).Info("Pod was evicted, but the node is not under memory pressure.")
+					return nil
+				}
+				log.V(4).Info("Pod might have been evited because node was under memory pressure")
+			} else {
+				// The pod was oomkilled
+				// Check if scaling already happened after this was oomkilled
+				recent := false
+				for i := range pod.Status.ContainerStatuses {
+					containerStatus := &pod.Status.ContainerStatuses[i]
+					if containerStatus.RestartCount > 0 &&
+						containerStatus.LastTerminationState.Terminated != nil &&
+						containerStatus.LastTerminationState.Terminated.Reason == "OOMKilled" &&
+						clone.Status.HvpaStatus.LastScaleTime != nil &&
+						containerStatus.LastTerminationState.Terminated.FinishedAt.After(clone.Status.HvpaStatus.LastScaleTime.Time) {
+
+						recent = true
+						break
+					}
+				}
+				if !recent {
+					log.V(4).Info("This is not  a recent oomkill. Return")
+					return nil
+				}
+			}
+
+			clone.Status.HvpaStatus.OverrideScaleUpStabilization = true
+
+			log.V(2).Info("Updating HVPA status to override last scale time", "HVPA", clone.Name)
+			err = client.Update(context.TODO(), clone)
+			if err != nil {
+				log.Error(err, "Error overrinding last scale time for", "HVPA", name)
+			}
+
+			return nil
+		}),
+	}
+
+	pred := OomkillPredicate{
+		Funcs: predicate.Funcs{
+			UpdateFunc: updateEventFunc,
+		},
+	}
+
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, eventHandler, pred)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
+// OomkillPredicate implements a oomkill predicate function
+type OomkillPredicate struct {
+	predicate.Funcs
+}
+
+type hvpaObj struct {
+	Name     string
+	Selector labels.Selector
+}
+
+var cachedNames map[string][]*hvpaObj
```

Comment:
Why not `var cachedNames map[string]map[string]*hvpaObj`? This could help avoid [this](https://github.com/gardener/hvpa-controller/pull/4/files#diff-bfa21ee442000583bc7aa6f53ec3660cR419) loop.